imap_data_access/file_validation.py

```diff
@@ -0,0 +1,264 @@
+"""Methods for managing and validating filenames and filepaths"""
+from __future__ import annotations
+
+import re
+
+from datetime import datetime
+from pathlib import Path
+
+VALID_INSTRUMENTS = {
+    "codice",
+    "glows",
+    "hit",
+    "hi-45",
+    "hi-90",
+    "idex",
+    "lo",
+    "mag",
+    "swapi",
+    "swe",
+    "ultra-45",
+    "ultra-90",
+}
+
+VALID_DATALEVELS = {"l0", "l1", "l1a", "l1b", "l1c", "l1d", "l2"}
+
+VALID_FILE_EXTENSION = {"pkts", "cdf"}
+
+FILENAME_CONVENTION = (
+    "<mission>_<instrument>_<datalevel>_<descriptor>_"
+    "<startdate>_<enddate>_<version>.<extension>"
+)
+
+
+def extract_filename_components(filename: str | Path):
+    """
+    Extracts all components from filename. Does not validate instrument or level.
+
+    Will return a dictionary with the following keys:
+    { instrument, datalevel, descriptor, startdate, enddate, version, extension }
+
+    If a match is not found, a ValueError will be raised.
+
+    Parameters
+    ----------
+    filename : Path or str
+        Path of dependency data.
+
+    Returns
+    -------
+    components : dict
+        Dictionary containing components.
+
+    """
+    pattern = (
+        r"^imap_"
+        r"(?P<instrument>[^_]+)_"
+        r"(?P<datalevel>[^_]+)_"
+        r"(?P<descriptor>[^_]+)_"
+        r"(?P<startdate>\d{8})_"
+        r"(?P<enddate>\d{8})_"
+        r"(?P<version>v\d{2}-\d{2})"
+        r"\.(cdf|pkts)$"
+    )
+    if isinstance(filename, Path):
+        filename = filename.name
+
+    match = re.match(pattern, filename)
+    if match is None:
+        raise ValueError(f"Filename {filename} does not match expected pattern")
+
+    components = match.groupdict()
+    components["extension"] = match.group(7)
+    components["mission"] = "imap"
+    return components
+
+
+def construct_upload_path(filename: str | Path) -> Path:
```

Comment:
I think this might be used in places other than just upload, so something more generic might be good in the name, `construct_file_path_from_name` (a little long, so not super ideal, but just an idea)